Add documentation for container checkpointing feature (KEP 2008) (#31753)
kubernetes/website: content/en/docs/reference/node/kubelet-checkpoint-api.md

---
content_type: "api_reference"
title: Kubetlet Checkpoint API
weight: 10
---

> ## Kubelet Checkpoint API

Reviewer: ```suggestion
```
The theme renders a page title for you.

> Checkpointing a container is the functionality to create a stateful copy of a
> running container. The stateful copy of the container can be moved to another
> system and after restore the container will continue to run at exactly the same
> point it was checkpointed. The concept of checkpointing a container and
> restoring a container can be seen as pausing and unpausing a container with the
> possibility to transfer the container to another system between pausing and
> unpausing.

Reviewer: OK, so now this is in place, I'll suggest wording tweaks:
```suggestion
Checkpointing a container is the functionality to create a stateful copy of a
running container. Once you have a stateful copy of a container, you could
move it to a different computer for debugging or similar purposes.

If you move the checkpointed container data to a computer that's able to restore
it, that restored container continues to run at exactly the same
point it was checkpointed. You can also inspect the saved data, provided that you
have suitable tools for doing so.
```

I don't want to give people the impression that restoring to a kubelet is supported - it isn't. And we [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future) in our docs.

## Operations {#Operations}

> ### `post` checkpoint the specified container

Reviewer: ```suggestion
### `post` checkpoint the specified container {#post-checkpoint}
```

> Tell the kubelet to checkpoint the specified container.

Reviewer: ```suggestion
Tell the kubelet to checkpoint a specific container from the specified Pod.
```

> 

Author: Added a line with a link to the Kubelet authentication/authorization reference

The kubelet will request a checkpoint from the underlying
{{<glossary_tooltip term_id="cri" text="CRI">}} implementation. In the checkpoint
request the kubelet will specify the name of the checkpoint archive as
`<containerID>.tar` and also request to store the checkpoint archive in the
`checkpoints` directory below its root directory (as defined by `--root-dir`).

> This defaults to `/var/lib/kubelet/checkpoints`.

Author: Added information that the checkpoint archive is overwritten and that it depends on the CRI implementation.

> The content of the checkpoint archive depends on the underlying CRI implementation.

Reviewer: ```suggestion
The checkpoint archive is in _tar_ format, and could be listed using an implementation of
[`tar`](https://pubs.opengroup.org/onlinepubs/7908799/xcu/tar.html). The contents of the
archive depend on the underlying CRI implementation (the container runtime on that node).
```

#### HTTP Request

POST /checkpoint/{namespace}/{pod}/{container}

#### Parameters

- **namespace** (*in path*): string, required

  {{< glossary_tooltip term_id="namespace" >}}

- **pod** (*in path*): string, required

  {{< glossary_tooltip term_id="pod" >}}

- **container** (*in path*): string, required

  {{< glossary_tooltip term_id="container" >}}

#### Response

200: OK

> 401: Unauthorized

Reviewer: What happens if
- the feature isn't enabled?
- the container runtime doesn't support checkpointing?
- the checkpoint archive already exists?
- there is no storage left?

For alpha, it might be OK to say that the kubelet can also return other statuses, and maybe put in an HTML comment to say that this should be filled out properly for beta.

Author: Also added more details.
